tensorflow-opgen/src/TensorFlow/OpGen.hs

```diff
@@ -237,40 +245,57 @@ quotedText = dquotes . strictText
 typeSig :: OpDef -> Doc
 typeSig d =
     foralls <+> constraints <+/>
-    signatureFold (mandatoryAttrInputs ++ tensorInputs ++ [outputs])
+    signatureFold (mandatoryAttrInputs ++ map snd tensorInputs ++ [outputs])
   where
-    foralls | Map.null typeMap = empty
+    foralls | null typeMap = empty
             | otherwise =
               "forall"
-              <+> sep (refTypes ++ map (strictText . fst) (Map.elems typeMap))
+              <+> sep (refVariableNames ++ typeMapTypeNames)
               <+> "."
-    constraints | Map.null typeMap = empty
+    typeMapTypeNames = map (strictText . fst) (Map.elems typeMap)
+    constraints | null typeMap = empty
                 | otherwise =
                   tuple (concatMap
                          (\(t, aDef) ->
                            "TensorType" <+> strictText t
                            : maybeToList (oneOfRestrictions aDef t))
                          (Map.elems typeMap)) <+> "=>"
     tensorInputs = zipWith tensorArg refTypes (d ^. inputArg)
-    refTypes = map (\x -> "v" <> int x) [1..length (d ^. inputArg)]
-    tensorArg refType arg = wrapArg refType arg <+>
-                            hang 0 ("-- ^" <+> argComment arg)
+    refTypes = ["v" <> int x | x <- [1..length (d ^. inputArg)]]
+    refVariableNames = catMaybes (map fst tensorInputs)
+    tensorArg refType arg = wrapArg refType arg <**>
+                            pure (<+> hang 0 ("-- ^" <+> argComment arg))
     -- Argument type is a list of tensors if number_attr is set;
     -- otherwise it's a single Tensor.
     wrapArg refType arg =
-        if Text.null (arg ^. numberAttr) then typ else brackets typ
+        if Text.null (arg ^. numberAttr) then typ else brackets <$> typ
       where typ = tensorType refType arg
-    tensorType refType arg =
-      "Tensor" <+> refType <+> maybe directType strictText indirectType
+    -- The result is (reference type variable if any, type representing the arg)
+    tensorType :: Doc -> OpDef'ArgDef  -> (Maybe Doc, Doc)
+    tensorType refType arg
+        -- Deals with resource handles that are unlike tensors and
+        -- have their own type level representation. The magic "dtype"
```

Comment:
It seems like this convention will break if there is an op that accepts two resource handles with different types.